xla/service/collective_ops_utils.cc

```diff
@@ -133,6 +133,31 @@ absl::StatusOr<std::vector<int>> GetParticipatingIDs(
                           group->replica_ids().end());
 }
 
+// Returns the group formation mode of instr, assuming that instr is, or is
+// dervied from, an HloAllGatherInstruction, HloAllReduceInstructionBase,
+// HloCollectiveInstruction or HloCollectivePermuteInstruction.
+absl::StatusOr<CollectiveOpGroupMode> GetCollectiveOpGroupMode(
+    HloInstruction* instr) {
+  if (HloAllGatherInstruction::ClassOf(instr)) {
+    auto collective = Cast<HloAllGatherInstruction>(instr);
+    return GetCollectiveOpGroupMode(collective->channel_id().has_value(),
+                                    collective->use_global_device_ids());
+  } else if (HloAllReduceInstructionBase::ClassOf(instr)) {
+    auto collective = Cast<HloAllReduceInstructionBase>(instr);
+    return GetCollectiveOpGroupMode(collective->channel_id().has_value(),
+                                    collective->use_global_device_ids());
+  } else if (HloCollectiveInstruction::ClassOf(instr)) {
```

Comment:
I'm worried if a new subclass is added with use_global_device_ids, we'll forget to update this. I would either handle each subclass of this separately, or check that it's a hardcoded list of known subclasses (all-to-all, ragged-all-to-all, collective-broadcast)